### DISCUSSION TITLE
Allow Java applications to be split from the VPN tunnel

### DISCUSSION THREAD

**Discussion by ghost** [mullvad/mullvadvpn-app#9134] [Feature requests & Ideas]
## Issue report

Operating system: Windows 11

App version: 2023.4

### Issue description

Whenever you add javaw.exe to split tunneling, you'd expect for javaw.exe to bypass mullvad, instead of blocks all connections from the process, even when windows firewall is disabled, giving a permissiondenied exception.

I know this has been reported in the past, but I am reporting again because it hasn't been fixed, I saw that the earliest occurrence of this reported had a "wont fix" label, which is incorrect, the issue must be fixed. No other VPNs have this problem.

**Comment by faern** [comment]
The java runtime initializes and handles socket in a very ugly non-standard way. This is currently incompatible with our split tunneling kernel driver. It might be possible, but would be pretty ugly to have the kernel driver accept and work around the hack that Java is doing. They basically only use IPv6 sockets, even for IPv4 connections.

This might be fixed in the future. But it's currently not a very high priority, sorry.

**Comment by AstudilloMarillo** [comment]
Just came to share my support for this feature request, would love it if compatibility for java applications would be implemented.

**Comment by FlowOverFail** [comment]
Would like to see this as well. Is there any work started on this that we could look at helping out with?

**Comment by MarkusPettersson98** [comment]
I'll convert this to a feature request, since Java applications are listed as explicitly unsupported by our split tunneling driver on Windows and is thus arguably not a bug: https://mullvad.net/en/help/split-tunneling-with-the-mullvad-app#unsupported. As mentioned, it might be possible to allow Java applications to be split by using a different implementation or rewriting the driver, and a big push from the community might motivate such a decision :)

**Comment by Tom1817** [comment]
I would love to see this, my partner and I recently started running an SMP and we both use mullvad as do several players

**Comment by SpoilerRules** [comment]
Please fix this.

**Comment by EzioisAwesome56** [comment]
a workaround for this is to add `-Djava.net.preferIPv4Stack=true` to your java launch arguments. This forces java (On windows atleast) to use ipv4 sockets, with the side effect of disabling ipv6 entirely.

It does appear that it allowed minecraft to function with split tunnel turned on though, so there's that
